pelican/generators.py

```diff
@@ -680,6 +634,74 @@ def _process(arts):
         self.dates.sort(key=attrgetter('date'),
                         reverse=self.context['NEWEST_FIRST_ARCHIVES'])
 
+        def _build_period_archives(sorted_articles):
+            period_archives = defaultdict(list)
```

Comment:
Agreed that it should be moved out of `generate_context`, but I'm not sure about `utils` as the destination, because its purpose is too specific to generators, specifically `ArticlesGenerator` even. How about making it a `@staticmethod` right after `generate_context`?

Reply:
Sure, it can be a method in `ArticleGenerator`, not necessarily a `@staticmethod` though. It can be an internal method and keep using the `self.settings` as is.

Reply:
I've pushed the changes for this. Didn't make it a static method, though I did turn settings into a parameter.